### ISSUE TITLE
Add a way to hard reset GDScript parsing/errors

### ISSUE THREAD

**Issue by MajorMcDoom** [godotengine/godot-proposals#9232]
### Describe the project you are working on

Any game

### Describe the problem or limitation you are having in your project

GDScript editor frequently gets into a state where errors are incorrectly reported based on some kind of outdated cached data, but the cache does not refresh until the editor is restarted. And sometimes it requires an even more drastic measure of deleting things from the `.godot` folder (I still have not figured out a reliable way to solve this issue every time).
![Screenshot 2024-03-04 101257](https://github.com/godotengine/godot-proposals/assets/13228932/893e2dd0-ff6b-4de7-9d5c-92e243ef1a71)
![Screenshot 2024-03-04 101224](https://github.com/godotengine/godot-proposals/assets/13228932/02c364d1-9c79-49d7-805c-642c358bdcfc)

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The behaviour described above is a bug, which should obviously be fixed. The proposed feature is a menu item in the script editor that allows the user to clear all cached data related to scripting (essentially as though you just cloned a new repo and opened up the project for the first time). This proposed feature aims to:
- Ease the pain if this bug takes a long time to solve
- Gives the user a quick way out of other current and future cache-related bugs

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

A menu item in the script editor, perhaps `File > Clear Script Cache Data`.
I think for this feature to work well, it has to not only clear _saved_ cache data, but also _runtime_ cache data.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No.

### Is there a reason why this should be core and not an add-on in the asset library?

It is core behaviour of the script editor, addressing a core issue.

**Comment by Calinou** [comment]
This is likely due to `.godot/global_script_class_cache.cfg` becoming out of sync with the named classes in the project.

I feel you shouldn't have to go through a menu every time it becomes out of sync, so we should look into resolving that particular issue instead: https://github.com/godotengine/godot/issues/78642

**Comment by MajorMcDoom** [comment]
Thanks for looking into it and providing links!

> This is likely due to `.godot/global_script_class_cache.cfg` becoming out of sync with the named classes in the project: [godotengine/godot#81867](https://github.com/godotengine/godot/issues/81867)
> 
> I feel you shouldn't have to go through a menu (or even press a keyboard shortcut) every time it becomes out of sync, so we should look into resolving that particular issue instead: [godotengine/godot#78642](https://github.com/godotengine/godot/issues/78642)

Be that as it may, there is definitely more than one bug which has similar symptoms. I've experienced these types of issues since 3.x, and they don't always fit a specific pattern. Of course you shouldn't _have to_ rely on a manual reset, and these bugs should definitely be fixed, but some kind of way out in those tricky situations is welcome. For example, you _shouldn't_ get corrupt scene files, but allow you to open them and fix them is still a good idea.

I would say it's always a good idea to have a way to clear cache of any sort, because there's always some possibility of it getting into a bad state, whether it's from a core bug or through a plugin or some kind of file system shenanigans.

The current state of things is that users already have to do their own hard reset anyway, and often not in the correct way. It's some combination of restarting the editor and deleting various files. At least if it were a menu option, it would be somewhat reliable. I don't think it should be bound to a keyboard shortcut, because it shouldn't be implied as a common operation.

> You may want to try this pull request: [godotengine/godot#86067](https://github.com/godotengine/godot/pull/86067)

This PR doesn't match my symptoms, and I'm experiencing my issues after a simple variable renaming, or duplicating classes, or moving things between folders. My experience has been quite unpredictable.

**Comment by MajorMcDoom** [comment]
Also I can say that I just deleted the cfg file, and it made my problem _worse_. Practically every script threw errors. But deleting the _entire_ `.godot` folder fixed it. So I believe there are multiple bugs at play here. But of course, that also means knowing which things to "reset" for this proposal would also be tricky.

**Comment by lostminds** [comment]
> This is likely due to `.godot/global_script_class_cache.cfg` becoming out of sync with the named classes in the project: [godotengine/godot#81867](https://github.com/godotengine/godot/issues/81867)

I've had a lot of these issues as well, often when changing a function signature (adding or removing a parameter) in a class, which doesn't seem to update correctly until I restart the editor. Since restarting works for me, having a reset would just be slightly more convenient. 

However, a benefit of having a reset/rebuild tool for the script cache could be to use as a debug tool, since such a tool could also compare the cache before and after rebuilding it to see what if anything changed and in this case what. If rebuilding does fix an issue this information showing what exactly had gotten out of sync and was changed when rebuilding might be very useful in tracking down and fixing these bugs so we won't need to manually rebuild the cache in the future.

**Comment by HolonProduction** [comment]
I don't think this proposal is actionable. There is no singular point of failure that we can just reset. GDScript issues often arise from complex interactions between different systems e.g. `ScriptServer`, `ResourceCache`, `GDScriptCache`. The exact stuff that would need recalculation varies for different issues, so the only actionable solution seems fixing the actual issues.
